include/madwf_ml.h

```diff
@@ -0,0 +1,372 @@
+#pragma once
+
+#include <vector>
+#include <random>
+#include <unordered_map>
+
+#include <color_spinor_field.h>
+#include <blas_quda.h>
+#include <madwf_transfer.h>
+#include <polynomial.h>
+
+#include <random_quda.h>
+#include <tune_quda.h>
+
+#include <device_vector.h>
+
+namespace quda
+{
+
+  struct MadwfAcc {
+
+    using transfer_float = float;
+
+    static constexpr madwf_ml::transfer_5D_type transfer_type = madwf_ml::transfer_5D_type::Spin;
+
+    // The parameters to be trained.
+    using device_container = device_vector<transfer_float>;
+    device_container device_param;
+
+    // The diagonal component to suppress the zero modes.
+    double mu;
+    int Ls_base;
+
+    int null_miniter;
+    double null_tol;
+    int train_maxiter;
+
+    // persistent buffers for reuse.
+    ColorSpinorField *forward_tmp = nullptr;
```

Comment:
You could keep @mathiaswagner happy and use `unique_ptr` here 😉

Reply:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b